lib/charms/loki_k8s/v0/loki_push_api.py

```diff
@@ -1553,9 +1554,16 @@ def __init__(
         )
         super().__init__(charm, relation_name, alert_rules_path, recursive)
         events = self._charm.on[relation_name]
+
         self.framework.observe(self._charm.on.upgrade_charm, self._on_logging_relation_changed)
         self.framework.observe(events.relation_changed, self._on_logging_relation_changed)
-        self.framework.observe(events.relation_departed, self._on_logging_relation_departed)
+
+        # This will keep track of broken relations so that we can ignore the endpoints
+        # we find in their application data bags
+        # will be reinitiated every time an external event
+        # (like relation broken) is emitted
+        self._ignored_relations = []  # type: list
```

Comment:
Interesting! Do we know for sure under what circumstances the charm is not reinitialized, for this to work?

Reply:
Charm is initialized for every hook coming from Juju. Events that are emitted by a library like this look like recursive calls and do NOT run the charm constructor again.

Reply:
More info provided in 1b933eafc48a5378e051621b8c7fae397a9b065c